Makefile

```diff
@@ -0,0 +1,270 @@
+# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+# Use bash as the shell when executing a rule's recipe. For more details:
+# https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html
+SHELL := bash
+
+# Set the package's name, version, and path for use throughout the Makefile.
+PACKAGE_NAME := macaron
+PACKAGE_VERSION := $(shell python -c $$'try: import $(PACKAGE_NAME); print($(PACKAGE_NAME).__version__);\nexcept: print("unknown");')
+MACARON_PATH := $(shell pwd)
+
+# This variable contains the first goal that matches any of the listed goals
+# here, else it contains an empty string. The net effect is to filter out
+# whether this current run of `make` requires a Python virtual environment
+# by checking if any of the given goals requires a virtual environment (all
+# except the 'venv' and the various 'clean' and 'nuke' goals do). Note that
+# checking for 'upgrade' and 'check' goals includes all of their variations.
+NEED_VENV := $(or \
+  $(findstring all,$(MAKECMDGOALS)), \
+  $(findstring setup,$(MAKECMDGOALS)), \
+  $(findstring upgrade,$(MAKECMDGOALS)), \
+  $(findstring sbom,$(MAKECMDGOALS)), \
+  $(findstring requirements,$(MAKECMDGOALS)), \
+  $(findstring audit,$(MAKECMDGOALS)), \
+  $(findstring check,$(MAKECMDGOALS)), \
+  $(findstring test,$(MAKECMDGOALS)), \
+  $(findstring dist,$(MAKECMDGOALS)), \
+  $(findstring docs,$(MAKECMDGOALS)), \
+  $(findstring prune,$(MAKECMDGOALS)), \
+)
+ifeq ($(NEED_VENV),)
+  # None of the current goals requires a virtual environment.
+else
+  ifeq ($(origin VIRTUAL_ENV),undefined)
+    $(warning No Python virtual environment found, proceeding anyway)
+  else
+    ifeq ($(wildcard .venv/upgraded-on),)
+      $(warning Python virtual environment not yet set up, proceeding anyway)
+    endif
+  endif
+endif
+
+# If the project configuration file has been updated (package deps or
+# otherwise) then warn the user and suggest resolving the conflict.
+ifeq ($(shell test pyproject.toml -nt .venv/upgraded-on; echo $$?),0)
```

Comment:
> When I run `make venv` the first time, this warning is issued. I think we should only issue when the virtual environment existed but `make setup` hasn't been run in the current directory?

When you run `make venv` two times without running `make setup`,  `.venv/upgraded-on` does not exist (because `make setup` has not run yet), Makefile concludes that`pyproject.toml` is newer than (`-nt`) `.venv/upgraded-on`, so we generate the error to guide the user.